pkg/kratos/service.go

```diff
@@ -156,6 +156,11 @@ func (s *Service) MustReAuthenticate(ctx context.Context, hydraLoginChallenge st
 		return false, nil
 	}
 
+	// Backup code was used for login, no need to re-auth
+	if validateHash(hydraLoginChallenge, c.LoginChallengeHash) && c.BackupCodeUsed {
```

Comment:
I think that this `if` can be merged with the one above, no need to calculate the hash 2 times.